pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+
+  class Traversable(Protocol):
+      leafs: Iterable['Traversable']
+
+
+Using Protocols
+===============
+
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
+protocol types. For variables and parameters with protocol types, subtyping
+relationships are subject to the following rules:
+
+* A protocol is never a subtype of a concrete type.
+* A concrete type or a protocol ``X`` is a subtype of another protocol ``P``
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
+* Edge case: for recursive protocols, structural subtyping is decided
+  positively for situations where such decision depends on itself. Continuing
+  the previous example::
+
+    class Tree(Generic[T]):
+        def __init__(self, value: T,
+                           leafs: 'List[Tree[T]]') -> None:
+            self.value = value
+            self.leafs = leafs
+
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
+
+Generic protocol types follow the same rules of variance as non-protocol
+types. Protocols can be used in all special type constructors provided
+by the ``typing`` module and follow the corresponding subtyping rules.
+For example, protocol ``PInt`` is a subtype of ``Union[PInt, int]``.
+
+
+``Union[]`` and ``All[]``
+-------------------------
+
+``Union`` of protocol classes behaves the same way as for non-protocol
+classes. For example::
+
+  from typing import Union, Optional, Protocol
+
+  class Exitable(Protocol):
+      @abstracrmethod
+      def exit(self) -> int:
+          ...
+  class Quitable(Protocol):
+      @abstractmethod
+      def quit(self) -> Optional[int]:
+          ...
+
+  def finish(task: Union[Exitable, Quitable]) -> int:
+      ...
+  class GoodJob:
+      ...
+      def quit(self) -> int:
+          return 0
+  finish(GoodJob()) # OK
+
+In addition, we propose to add another special type construct
+``All`` that represents intersection types. Although for normal types
+it is not very useful, there are many situations where a variable should
+implement more than one protocol. Annotation by ``All[Proto1, Proto2, ...]``
+means that a given variable or parameter must implement all protocols
+``Proto1``, ``Proto2``, etc. either implicitly or explicitly. Example::
+
+  from typing import Sequence, Hashable, All
+
+  def cached_func(args: All[Sequence[float], Hashable]) -> float:
+      ...
+  cached_func((1, 2, 3)) # OK, tuple is hashable and sequence
+
+The interaction between union and intersection types is specified by PEP 483,
+and basically reflects the corresponding interactions for sets.
+
+
+``Type[]`` with protocols
+-------------------------
+
+Variables and parameters annotated with ``Type[Proto]`` accept only concrete
```

Comment:
I think it could work like this for all ABCs. But it does not work like this in ``mypy `` now, and it is not specified in PEP 484. I feel like this should be specified here rather that in PEP 484.

By the way, there is an old high-priority issue for this https://github.com/python/mypy/issues/1843 and my PR https://github.com/python/mypy/pull/2853 that didn't receive a review for a month :-)